File: 543/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        length(root);
        return ans;
    }
    int ans=0;
    int length(TreeNode root) {
        if(root==null) return 0; // or -1
        int right = length(root.right);
        int left = length(root.left);
        if(right+left>ans) ans = right + left;
        
        if(right>left) return right+1;
        else return left+1;
    }
}

File: 543/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# 想長出 diameter, 就把 length = depth(root.left)+depth(root.right)+2 找最長的 length
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxLength = 0
        def depth(root):
            if root==None:
                return 0 # 提早結束的終止條件
                
            left, right = 0, 0
            if root.left != None:
                left = depth(root.left) + 1
            if root.right != None:
                right = depth(root.right) + 1
            length = left + right
            if length > self.maxLength:
                self.maxLength = length
            return max(left, right)
        
        depth(root) # 整個 tree 都測過一次
        return self.maxLength
        
